### ISSUE TITLE
Too many iptables entries being created

### ISSUE THREAD

**Issue by SerialVelocity** [cloudnativelabs/kube-router#896]
This might be related to #795 as this causes memory usage to balloon.

I have a large amount of `KUBE-POD-FW-*` and `KUBE-NWPLCY-*` rules being generated that shouldn't be there:
```
-A KUBE-NWPLCY-JETO5HCCICPORJFD -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
-A KUBE-NWPLCY-JMA3X7HLSBOHV3YX -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
-A KUBE-NWPLCY-K6U2HGLSIC4TKVPJ -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
-A KUBE-NWPLCY-KVM3JJIFA3NGXEWT -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
-A KUBE-NWPLCY-KWBSGSZXTYUTZETX -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
-A KUBE-NWPLCY-KYO7N7MAAFNMMDPD -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
-A KUBE-NWPLCY-L3ZJ76FS4H34WKAT -p tcp -m comment --comment "rule to ACCEPT traffic from source pods to dest pods selected by policy name mysql namespace other" -m set --match-set KUBE-SRC-RKCNJAY4ZMHT3ECQ src -m set --match-set KUBE-DST-VAWT2357PZRYVGDI dst -m tcp --dport 3306 -j ACCEPT
```

I currently have 122 iptables entries that look just like that and then many more pointing to a different `KUBE-DST-*`.

```
$ sudo iptables-save | wc -l
16685
$ sudo iptables-save | grep KUBE-NWPLCY | wc -l
9874
$ kubectl get pods --all-namespaces | wc -l
155
$ kubectl get networkpolicies --all-namespaces | wc -l
11
```

On a host that has just started up:
```
$ sudo iptables-save | grep KUBE-NWPLCY | wc -l
39
```

Each network policy is small too.

I was using the `cloudnativelabs/kube-router:latest` docker image with flags:
```
        - "--run-router=false"
        - "--run-firewall=true"
        - "--run-service-proxy=false"
        - "--metrics-port=20249"
```

Downgrading to `cloudnativelabs/kube-router:v0.4.0` seems to have fixed it.

I do not currently have logs but I seem to remember them being empty?

**Comment by murali-reddy** [comment]
> I do not currently have logs but I seem to remember them being empty?

@SerialVelocity can you please enable verbose logging (--v=3)? Stale iptables chain representing a network policy are deleted during periodic  syncs. It must be failing to delete iptables chain. Verbose logs might have some details.

**Comment by salanki** [comment]
I run 0.4.0. When I disable Firewall my memory usage drops to ~200M and is steady. I run Kube-proxy in iptables mode, I wonder if that can play in. Do you run Kube-proxy with IPVS @SerialVelocity?

**Comment by SerialVelocity** [comment]
I use kube-proxy with iptables.

**Comment by SerialVelocity** [comment]
@murali-reddy I just tried the `latest` tag again and I see this in the logs:
```
E0508 12:30:04.023408       1 network_policy_controller.go:186] Error syncing network policy for the update to pod: namespace/jobname-1588941000-wjspk Error: Aborting sync. Failed to cleanup stale iptables rules: Failed to delete the chain KUBE-POD-FW-BVDBJK7GATSVEVMT due to running [/sbin/iptables -t filter -X KUBE-POD-FW-BVDBJK7GATSVEVMT --wait]: exit status 1: iptables: Too many links.
```

**Comment by murali-reddy** [comment]
> Error syncing network policy for the update to pod: namespace/jobname-1588941000-wjspk Error: Aborting sync. Failed to cleanup stale iptables rules: Failed to delete the chain KUBE-POD-FW-BVDBJK7GATSVEVMT due to running [/sbin/iptables -t filter -X KUBE-POD-FW-BVDBJK7GATSVEVMT --wait]: exit status 1: iptables: Too many links

it looks like kube-router hit a fatal situation from where it can not recover. To overcome if you flush iptables `filter` table it would come to normal working state. Let me look at the cleanup logic in kube-router on how it can recover and avoid excessive growing number of chains

**Comment by aauren** [comment]
I can confirm that we're experiencing this on a cluster after upgrading to 1.0.0-rc2. We did not see this behavior with 1.0.0-rc1, so the bug must be somewhere within that changeset.

**Comment by aauren** [comment]
I did a test by taking a node down to a single pod and building a new kube-worker binary with and without #891 and if I revert #891 kube-router functionality returns to normal in terms of iptables entries created.

So I'm guessing that our cleanup code doesn't handle the addition of linking to policy chains from the OUTPUT tables correctly.

**Comment by aauren** [comment]
This should be fixed in v1.0.0-rc3. The containers should be on docker hub right now. Please let us know if you see any further issues with iptables growing unchecked with the new release.
